### PR TITLE
refactor: parsing with remaining input

### PR DESCRIPTION
**Pull request by senia-psm**
new method parse_bytes in FromBytes trait
parse_bytes returns result with remaining input

### CODE REVIEW

**Comment by suhr** [comment]
(read first lines so far)

> `ParseResult`

This type is basically a copy of Result/Option. I don't think this is a good idea.
Better to introduce a proper `ParseError` type.

**Comment by senia-psm** [comment]
@suhr, I can't see any good options for `ParseError` now.
The goal was to add remaining input to fix this: https://github.com/zetok/tox/blob/master/src/toxcore/dht.rs#L716
And to reduce amount of constants in code (call side should not know the length of input bytes for called method).
I can replace `ParseResult` with this:

``` rust
struct Parsed<'a, Output>{out: Output, result: &'a [u8]}
...
fn parse_bytes(bytes: &[u8]) -> Option<Parsed<Output>>;
```

**Comment by suhr** [comment]
> I can't see any good options for `ParseError` now.

Well, type `Result<Parsed<...>, ParseError>` have clearer semantics than `Option<Parsed>`. And you can move all that `debug!("Not enough bytes for Ipv6Addr!");` into it.

> `Parsed<'a, Output>`

This makes sense. Also you can do

``` rust
pub type ParseResult<'a, Output> = Result<Parsed<'a, Output>, ParseError>
```

Not a new type, but a type alias.

**Comment by senia-psm** [comment]
@suhr, we'll need an implementation of `ValidationNel` to get read of `debug!`. And it's quite impossible to create a good implementation of `ValidationNel` in rust (without higher kinded polymorphism). So I'll stick with `pub struct Parsed<'a, Output>(Output, &'a [u8]);` and `Option<Parsed<Output>>`.

I guess tuple struct is better here wince we have to pattern match over it every time.

What do you think about this solution?

**Comment by suhr** [comment]
I guess you should look how it's done in rust:
https://doc.rust-lang.org/book/error-handling.html
https://doc.rust-lang.org/std/io/type.Result.html

Rust idioms are different from Scala's one. And it is idiomatic to use `Result` for values with errors.

> I guess tuple struct is better here wince we have to pattern match over it every time.

Turple structs are fine.

**Comment by senia-psm** [comment]
@suhr , please review new version with `Result` and `try!`.

File: src/toxcore/dht.rs [hunk 1]
```diff
@@ -593,24 +541,23 @@ impl ToBytes for GetNodes {
...
+            return parse_error!("Amount of bytes is less than GET_NODES_SIZE!")
         }
 
         if let Some(pk) = PublicKey::from_slice(&bytes[..PUBLICKEYBYTES]) {
             // need shorter name for ID bytes
             let b = &bytes[PUBLICKEYBYTES..GET_NODES_SIZE];
             let id = array_to_u64(&[b[0], b[1], b[2], b[3],
                                     b[4], b[5], b[6], b[7]]);
-            return Some(GetNodes { pk: pk, id: id })
+            Ok(Parsed(GetNodes { pk: pk, id: id }, &bytes[PUBLICKEYBYTES..]))
```
Comment (zetok) [inline src/toxcore/dht.rs]: Shouldn't be `GET_NODES_SIZE` ?

**Comment by suhr** [comment]
Looks fine. I'd rather use enums (a lot of errors are about `NotEnoughBytes(for_thing)` and `Invalid(something)`) and implement `Display` trait (and maybe `Error`) instead of using raw strings, but this is a minor thing that can be added later.

File: src/toxcore/dht.rs [hunk 2]
```diff
@@ -945,59 +877,46 @@ impl ToBytes for DhtPacket {
...
 
-        const NONCE_POS: usize = 1 + PUBLICKEYBYTES;
-        let sender_pk = match PublicKey::from_slice(&bytes[1..NONCE_POS]) {
+        let sender_pk = match PublicKey::from_slice(&bytes[..PUBLICKEYBYTES]) {
             Some(pk) => pk,
             None => {
-                debug!("Failed; de-serializing sender's PK!");
-                trace!("With bytes for PK: {:?}", &bytes[1..NONCE_POS]);
-                return None
+                return parse_error!("Failed; de-serializing sender's PK! With bytes for PK: {:?}", &bytes[..PUBLICKEYBYTES])
```
Comment (zetok) [inline src/toxcore/dht.rs]: could be split into 2 lines

File: src/toxcore/dht.rs [hunk 3]
```diff
@@ -945,59 +877,46 @@ impl ToBytes for DhtPacket {
...
 
-        const PAYLOAD_POS: usize = NONCE_POS + NONCEBYTES;
-        let nonce = match Nonce::from_slice(&bytes[NONCE_POS..PAYLOAD_POS]) {
+        let nonce = match Nonce::from_slice(&bytes[..NONCEBYTES]) {
             Some(n) => n,
             None => {
-                debug!("Failed; de-serializing nonce!");
-                trace!("With bytes for nonce: {:?}", &bytes[NONCE_POS..PAYLOAD_POS]);
-                return None
+                return parse_error!("Failed; de-serializing nonce! With bytes for nonce: {:?}", &bytes[..NONCEBYTES])
```
Comment (zetok) [inline src/toxcore/dht.rs]: also could be split

File: src/toxcore/state_format/old.rs [hunk 4]
```diff
@@ -597,68 +614,52 @@ pub const FRIENDSTATEBYTES: usize = 1      // "Status"
...
-        // supply start position and *length*
-        let get_string = |start, len: usize| -> Option<String> {
+        // parse string out of bytes
+        // supply length
+        fn parse_string(bytes: &[u8], len: usize) -> ParseResult<String> {
             let str_len = u16::from_be(array_to_u16(
-                        &[bytes[start+len], bytes[start+len+1]])) as usize;
-            String::from_utf8(bytes[start..start+str_len].to_vec()).ok()
+                        &[bytes[len], bytes[len+1]])) as usize;
+            match String::from_utf8(bytes[..str_len].to_vec()).ok() {
```
Comment (zetok) [inline src/toxcore/state_format/old.rs]: perhaps `ok()` could be removed? Without it, `Err(e)` could provide more useful info for `parse_error!`

File: src/toxcore_tests/dht_tests.rs [hunk 5]
```diff
@@ -344,50 +391,90 @@ fn packed_node_ip_test() {
...
-        let mut bytes = vec![];
-        for n in nodes.clone() {
-            bytes.extend_from_slice(&n.to_bytes());
+fn packed_node_parse_bytes_multiple_n_test() {
+    fn with_nodes(nodes: Vec<PackedNode>, random_rest: Vec<u8>) {
+        // should parse nodes
+        {
+            let mut bytes = vec![];
+            for n in nodes.clone() {
+                bytes.extend_from_slice(&n.to_bytes());
```
Comment (zetok) [inline src/toxcore_tests/dht_tests.rs:498]: without unneeded `.clone()`:

``` rust
for n in &nodes {
    bytes.extend_from_slice(&n.to_bytes());
```

File: src/toxcore_tests/dht_tests.rs [hunk 6]
```diff
@@ -1340,6 +1493,18 @@ fn nat_ping_from_bytes_test() {
     with_bytes(ping);
 }
 
+// NatPing::parse_bytes()
+
+#[test]
+fn nat_ping_parse_bytes_rest_test() {
+    fn with_bytes(bytes: Vec<u8>) {
+        if let Ok(Parsed(_, rest)) = NatPing::parse_bytes(&bytes) {
```
Comment (zetok) [inline src/toxcore_tests/dht_tests.rs]: wouldn't it be better to use `NatPing` as function arg, `.to_bytes()` it append some random bytes & test that?

File: src/toxcore_tests/packet_kind_tests.rs [hunk 7]
```diff
@@ -64,3 +66,15 @@ fn packet_kind_from_bytes_test() {
...
 }
+
+// PacketKind::parse_bytes
+
+#[test]
+fn packet_kind_parse_bytes_rest_test() {
+    fn with_bytes(bytes: Vec<u8>) {
+        if let Ok(Parsed(_, rest)) = PacketKind::parse_bytes(&bytes) {
+            assert_eq!(&bytes[1..], rest);
+        }
```
Comment (zetok) [inline src/toxcore_tests/packet_kind_tests.rs]: `if` branch should return `TestResult::passed()`, and there should be added `else` branch that would return `TestResult::discard()`; at this point though `match` would be better.

e.g. https://github.com/zetok/tox/blob/771ef573e620ad672f53d1c801e16bdc3f506760/src/toxencryptsave_tests/encryptsave_tests.rs#L50,L63

**Pull request merged**